flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -2527,6 +2527,39 @@ class ScalarArrayExprLowering {
     return lexv;
   }
 
+  ExtValue userAssign(mlir::FuncOp userAssignment,
+                      const Fortran::lower::SomeExpr &lhs,
+                      const Fortran::lower::SomeExpr &rhs) {
+    semant = ConstituentSemantics::RefTransparent;
+    // 1) Lower the rhs expression with array_fetch op(s).
+    auto rexv = lower(rhs);
+    elementalValue = fir::getBase(rexv);
+    // 2) Lower the lhs expression to an array_update.
+    semant = ConstituentSemantics::CustomCopyInCopyOut;
+    auto lexv = lower(lhs);
+    bool isIllFormedLHS = false;
+    // 3) Insert the call
+    if (auto modifyOp = mlir::dyn_cast<fir::ArrayModifyOp>(
+            fir::getBase(lexv).getDefiningOp())) {
+      auto oldInnerArg = modifyOp.sequence();
+      auto offset = expSpace.argPosition(oldInnerArg);
+      expSpace.setInnerArg(offset, fir::getBase(lexv));
+      genScalarUserDefinedAssignmentCall(builder, getLoc(), userAssignment,
+                                         modifyOp.getResult(0), elementalValue);
+    } else {
+      // Ill-formed lhs. Still insert user assignment call for consistency.
```

Comment:
This sounds bugly. Should we just call `fir::emitFatalError` here?

Reply:
The normal assignment case was updated in [this commit](https://github.com/flang-compiler/f18-llvm-project/commit/0648a8186f56041b6a8fb573aa0f3c59d2cfe9e0) to support degenerated assignment. I do not see why user assignment should not also support it if this was needed for normal assignment.

Reply:
I see. Can we update these comments to indicate that we're performing a scalar assignment in an array expression context? This violates the standard, I believe, but I agree we might as well allow it everywhere.

Reply:
Sure, I have updated the comment.